File: langs/python/src/vmx/collections/observable_list.py
"""ObservableList — granular per-mutation event observable list.

See spec/21-collections.md §3 and ADR-0026.
"""

from __future__ import annotations

from collections.abc import Generator, Iterable, Iterator
from contextlib import contextmanager
from typing import Generic, TypeVar, overload

import reactivex as rx
from reactivex import operators as ops
from reactivex.subject import Subject

T = TypeVar("T")


class ObservableList(Generic[T]):
    """An observable list that raises granular per-mutation events.

    Four observables are exposed:

    - ``on_item_added``    — ``Observable[tuple[T, int]]`` emits ``(item, index)``
    - ``on_item_removed``  — ``Observable[tuple[T, int]]`` emits ``(item, index)``
      (index is position before removal)
    - ``on_item_replaced`` — ``Observable[tuple[T, T, int]]`` emits ``(new_item, old_item, index)``
    - ``on_reset``         — ``Observable[None]``

    A ``PropertyChanged("Count")`` signal is emitted via ``on_property_changed``
    **after** the granular event for every mutation that changes Count (add and
    remove; not replace). This ordering is normative per spec §3.3 and ADR-0026.

    Inside a ``batch_update()`` context manager, granular events are suppressed.
    A single ``Reset`` fires when the outermost batch exits (ref-counted).

    See spec/21-collections.md §3 and ADR-0026.
    """

    def __init__(self) -> None:
        self._items: list[T] = []
        self._batch_depth: int = 0
        self._mutated_in_batch: bool = False
        self._count_at_batch_start: int = 0
        self._disposed: bool = False

        # Granular event subjects
        self._added_subject: Subject[tuple[T, int]] = Subject()
        self._removed_subject: Subject[tuple[T, int]] = Subject()
        self._replaced_subject: Subject[tuple[T, T, int]] = Subject()
        self._reset_subject: Subject[None] = Subject()
        # Property-changed subject (emits the property name as a string)
        self._prop_changed_subject: Subject[str] = Subject()

    # ── Observable properties ─────────────────────────────────────────────────
    # Each Subject is sealed behind ``as_observable`` so external subscribers
    # can only subscribe — never ``on_next``/``dispose`` the internal stream and
    # corrupt other subscribers (VMX-013).

    @property
    def on_item_added(self) -> rx.Observable[tuple[T, int]]:
        """Hot observable: emits ``(item, index)`` on every add/insert."""
        return self._added_subject.pipe(ops.as_observable())

    @property
    def on_item_removed(self) -> rx.Observable[tuple[T, int]]:
        """Hot observable: emits ``(item, index)`` (index before removal) on remove."""
        return self._removed_subject.pipe(ops.as_observable())

    @property
    def on_item_replaced(self) -> rx.Observable[tuple[T, T, int]]:
        """Hot observable: emits ``(new_item, old_item, index)`` on replace."""
        return self._replaced_subject.pipe(ops.as_observable())

    @property
    def on_reset(self) -> rx.Observable[None]:
        """Hot observable: emits ``None`` on clear or batch completion."""
        return self._reset_subject.pipe(ops.as_observable())

    @property
    def on_property_changed(self) -> rx.Observable[str]:
        """Hot observable: emits the property name on every property change."""
        return self._prop_changed_subject.pipe(ops.as_observable())

    # ── Count / indexing ──────────────────────────────────────────────────────

    @property
    def count(self) -> int:
        """Number of items in the list."""
        return len(self._items)

    def __len__(self) -> int:
        return len(self._items)

    def __iter__(self) -> Iterator[T]:
        return iter(self._items)

    @overload
    def __getitem__(self, index: int) -> T: ...

    @overload
    def __getitem__(self, index: slice) -> list[T]: ...

    def __getitem__(self, index: int | slice) -> T | list[T]:
        if isinstance(index, slice):
            return self._items[index]
        return self._items[index]

    # ── Mutations ─────────────────────────────────────────────────────────────

    def add(self, item: T) -> None:
        """Append *item* to the end of the list (alias for :meth:`append`)."""
        self.append(item)

    def append(self, item: T) -> None:
        """Append *item* to the end of the list."""
        index = len(self._items)
        self._items.append(item)
        self._on_added(item, index)

    def insert(self, index: int, item: T) -> None:
        """Insert *item* at *index* (stdlib semantics: negative indexes count
        from the end; out-of-range indexes clamp like :meth:`list.insert`).
        """
        # Normalize before emitting: spec/21 §3.2 mandates the payload carry
        # the actual insertion index — a raw -1 (or an out-of-range 99 that
        # stdlib silently clamps to len) violates the contract.
        if index < 0:
            index = max(index + len(self._items), 0)
        elif index > len(self._items):
            index = len(self._items)
        self._items.insert(index, item)
        self._on_added(item, index)

    def remove(self, item: T) -> bool:
        """Remove the first occurrence of *item*. Returns ``True`` if removed."""
        try:
            idx = self._items.index(item)
        except ValueError:
            return False
        del self._items[idx]
        self._on_removed(item, idx)
        return True

    def remove_at(self, index: int) -> None:
        """Remove the item at *index* (negative indexes count from the end)."""
        # Normalize before emitting: spec/21 §3.2 mandates the payload carry
        # the index *before removal* — a raw -1 violates the contract even
        # though negative indexing itself is Python-idiomatic (ADR-0006).
        if index < 0:
            index += len(self._items)
            if index < 0:
                raise IndexError("list index out of range")
        item = self._items[index]
        del self._items[index]
        self._on_removed(item, index)

    def replace(self, index: int, new_item: T) -> None:
        """Replace the item at *index* (negative indexes count from the end)."""
        if index < 0:
            index += len(self._items)
            if index < 0:
                raise IndexError("list index out of range")
        old_item = self._items[index]
        self._items[index] = new_item
        self._on_replaced(new_item, old_item, index)

    def replace_all(self, items: Iterable[T]) -> None:
        """Replace all contents from a snapshot and emit one reset.

        Empty-to-empty is the only no-op. Equal-length and element-for-element
        identical non-empty inputs still emit Reset without comparing elements.
        """
        snapshot = list(items)
        old_count = len(self._items)
        if old_count == 0 and not snapshot:
            return

        self._items[:] = snapshot
        self._on_reset()
        if old_count != len(snapshot) and self._batch_depth == 0 and not self._disposed:
            self._prop_changed_subject.on_next("Count")

    def clear(self) -> None:
        """Remove all items, emitting Reset then ``Count`` — but only when the
        list was non-empty. Clearing an already-empty list changes nothing and
        emits nothing (ADR-0037 §2.2, mirroring the empty-batch precedent)."""
        count_changed = bool(self._items)
        self._items.clear()
        # Clearing an empty list is a no-op: emit neither Reset nor Count.
        if count_changed:
            self._on_reset()
        # spec/21 §3.3: PropertyChanged("Count") fires after every mutation
        # that changes Count. Inside a batch the batch-exit path emits it.
        if count_changed and self._batch_depth == 0 and not self._disposed:
            self._prop_changed_subject.on_next("Count")

    # ── Batch update ──────────────────────────────────────────────────────────

    @contextmanager
    def batch_update(self) -> Generator[None, None, None]:
        """Context manager that suppresses granular events during the block.

        On exit of the outermost batch (ref-counted), a single ``Reset`` is
        emitted if any mutations occurred. If the count actually changed during
        the batch, a ``PropertyChanged("Count")`` is also emitted (spec §3.3).
        """
        if self._batch_depth == 0:
            self._count_at_batch_start = len(self._items)
        self._batch_depth += 1
        try:
            yield
        finally:
            self._batch_depth -= 1
            if self._batch_depth == 0 and self._mutated_in_batch and not self._disposed:
                final_count = len(self._items)
                self._mutated_in_batch = False
                self._reset_subject.on_next(None)
                # Emit Count notification only if count actually changed (spec §3.3).
                if final_count != self._count_at_batch_start:
                    self._prop_changed_subject.on_next("Count")

    # ── Disposal ──────────────────────────────────────────────────────────────

    def dispose(self) -> None:
        """Complete and dispose every backing Subject. Idempotent (VMX-096).

        After disposal the list no longer emits; subscribing to the (now
        completed) streams behaves like any disposed reactivex Subject.
        """
        if self._disposed:
            return
        self._disposed = True
        for subject in (
            self._added_subject,
            self._removed_subject,
            self._replaced_subject,
            self._reset_subject,
            self._prop_changed_subject,
        ):
            subject.on_completed()
            subject.dispose()

    # ── Internal helpers ──────────────────────────────────────────────────────

    def _on_added(self, item: T, index: int) -> None:
        if self._disposed:
            return
        if self._batch_depth > 0:
            self._mutated_in_batch = True
            return
        self._added_subject.on_next((item, index))
        self._prop_changed_subject.on_next("Count")

    def _on_removed(self, item: T, index: int) -> None:
        if self._disposed:
            return
        if self._batch_depth > 0:
            self._mutated_in_batch = True
            return
        self._removed_subject.on_next((item, index))
        self._prop_changed_subject.on_next("Count")

    def _on_replaced(self, new_item: T, old_item: T, index: int) -> None:
        if self._disposed:
            return
        if self._batch_depth > 0:
            self._mutated_in_batch = True
            return
        self._replaced_subject.on_next((new_item, old_item, index))
        # Count does not change on replace — no PropertyChanged("Count")

    def _on_reset(self) -> None:
        if self._disposed:
            return
        if self._batch_depth > 0:
            self._mutated_in_batch = True
            return
        self._reset_subject.on_next(None)


class ReadOnlyObservableList(Generic[T]):
    """Read-only facade over an :class:`ObservableList`.

    Exposes the query surface (``count``, length, iteration, indexing,
    containment) and the granular observable streams, but **none** of the
    mutation methods. A consumer handed this view can observe the list but
    cannot mutate it — so it cannot desync state the owner maintains in lockstep
    (VMX-014). The view reads through to the live source, so it always reflects
    the current contents.
    """

    __slots__ = ("_source",)

    def __init__(self, source: ObservableList[T]) -> None:
        self._source = source

    # ── Query surface ─────────────────────────────────────────────────────────

    @property
    def count(self) -> int:
        """Number of items currently in the underlying list."""
        return self._source.count

    def __len__(self) -> int:
        return len(self._source)

    def __iter__(self) -> Iterator[T]:
        return iter(self._source)

    @overload
    def __getitem__(self, index: int) -> T: ...

    @overload
    def __getitem__(self, index: slice) -> list[T]: ...

    def __getitem__(self, index: int | slice) -> T | list[T]:
        return self._source[index]

    def __contains__(self, item: object) -> bool:
        return any(item == existing for existing in self._source)

    # ── Observable surface (delegated; already sealed by the source) ───────────

    @property
    def on_item_added(self) -> rx.Observable[tuple[T, int]]:
        """Hot observable: emits ``(item, index)`` on every add/insert."""
        return self._source.on_item_added

    @property
    def on_item_removed(self) -> rx.Observable[tuple[T, int]]:
        """Hot observable: emits ``(item, index)`` on every remove."""
        return self._source.on_item_removed

    @property
    def on_item_replaced(self) -> rx.Observable[tuple[T, T, int]]:
        """Hot observable: emits ``(new_item, old_item, index)`` on replace."""
        return self._source.on_item_replaced

    @property
    def on_reset(self) -> rx.Observable[None]:
        """Hot observable: emits ``None`` on clear or batch completion."""
        return self._source.on_reset

    @property
    def on_property_changed(self) -> rx.Observable[str]:
        """Hot observable: emits the property name on every property change."""
        return self._source.on_property_changed

File: langs/csharp/src/VMx/Collections/ObservableList.cs
using System.Collections;
using System.Collections.Specialized;
using System.ComponentModel;

namespace VMx.Collections;

// ── Event-arg types ───────────────────────────────────────────────────────────

/// <summary>Event arguments for <see cref="ObservableList{T}.ItemAdded"/>.</summary>
public sealed class ItemAddedEventArgs<T> : EventArgs
{
    /// <summary>The item that was added.</summary>
    public T Item { get; }

    /// <summary>The index at which the item was inserted.</summary>
    public int Index { get; }

    internal ItemAddedEventArgs(T item, int index) { Item = item; Index = index; }
}

/// <summary>Event arguments for <see cref="ObservableList{T}.ItemRemoved"/>.</summary>
public sealed class ItemRemovedEventArgs<T> : EventArgs
{
    /// <summary>The item that was removed.</summary>
    public T Item { get; }

    /// <summary>The index of the item before it was removed.</summary>
    public int Index { get; }

    internal ItemRemovedEventArgs(T item, int index) { Item = item; Index = index; }
}

/// <summary>Event arguments for <see cref="ObservableList{T}.ItemReplaced"/>.</summary>
public sealed class ItemReplacedEventArgs<T> : EventArgs
{
    /// <summary>The new item at <see cref="Index"/>.</summary>
    public T NewItem { get; }

    /// <summary>The old item that was replaced.</summary>
    public T OldItem { get; }

    /// <summary>The index of the replaced item.</summary>
    public int Index { get; }

    internal ItemReplacedEventArgs(T newItem, T oldItem, int index)
    {
        NewItem = newItem;
        OldItem = oldItem;
        Index = index;
    }
}

// ── ObservableList<T> ─────────────────────────────────────────────────────────

/// <summary>
/// An observable list that raises granular per-mutation events:
/// <see cref="ItemAdded"/>, <see cref="ItemRemoved"/>, <see cref="ItemReplaced"/>,
/// and <see cref="Reset"/>.
///
/// Also implements <see cref="INotifyCollectionChanged"/> and
/// <see cref="INotifyPropertyChanged"/> for platform-binding compatibility
/// (spec §3.4 and ADR-0026 §3 rule 2).
///
/// <para>
/// The <see cref="Count"/> property change notification fires <b>after</b> the
/// granular event on every mutation that changes Count (add/remove; not replace).
/// This ordering is normative per spec §3.3 and ADR-0026.
/// </para>
///
/// <para>
/// Call <see cref="BatchUpdate"/> to suppress granular events; a single
/// <see cref="Reset"/> fires when the returned <see cref="IDisposable"/> is
/// disposed (ref-counted for nested scopes).
/// </para>
///
/// See spec/21-collections.md §3 and ADR-0026.
/// </summary>
/// <typeparam name="T">Element type.</typeparam>
public sealed class ObservableList<T> :
    IReadOnlyList<T>,
    INotifyCollectionChanged,
    INotifyPropertyChanged
{
    private readonly List<T> _items = [];
    private int _batchDepth;
    private bool _mutatedInBatch;
    private int _countAtBatchStart;

    // ── Granular events ───────────────────────────────────────────────────────

    /// <summary>Fires when an item is added. Payload: item and insertion index.</summary>
    public event EventHandler<ItemAddedEventArgs<T>>? ItemAdded;

    /// <summary>Fires when an item is removed. Payload: item and index before removal.</summary>
    public event EventHandler<ItemRemovedEventArgs<T>>? ItemRemoved;

    /// <summary>Fires when an item is replaced. Payload: new item, old item, and index.</summary>
    public event EventHandler<ItemReplacedEventArgs<T>>? ItemReplaced;

    /// <summary>Fires on <see cref="Clear"/> or when a batch with mutations completes.</summary>
    public event EventHandler? Reset;

    // ── Platform compatibility ────────────────────────────────────────────────

    /// <inheritdoc/>
    public event NotifyCollectionChangedEventHandler? CollectionChanged;

    /// <inheritdoc/>
    public event PropertyChangedEventHandler? PropertyChanged;

    // ── IReadOnlyList<T> ──────────────────────────────────────────────────────

    /// <inheritdoc/>
    public int Count => _items.Count;

    /// <inheritdoc/>
    public T this[int index] => _items[index];

    /// <inheritdoc/>
    public IEnumerator<T> GetEnumerator() => _items.GetEnumerator();

    IEnumerator IEnumerable.GetEnumerator() => _items.GetEnumerator();

    // ── Mutations ─────────────────────────────────────────────────────────────

    /// <summary>Append <paramref name="item"/> to the end of the list.</summary>
    public void Add(T item)
    {
        int index = _items.Count;
        _items.Add(item);
        OnAdded(item, index);
    }

    /// <summary>Insert <paramref name="item"/> at <paramref name="index"/>.</summary>
    public void Insert(int index, T item)
    {
        _items.Insert(index, item);
        OnAdded(item, index);
    }

    /// <summary>
    /// Remove the first occurrence of <paramref name="item"/>.
    /// Returns <see langword="true"/> if found and removed.
    /// </summary>
    public bool Remove(T item)
    {
        int idx = _items.IndexOf(item);
        if (idx < 0) return false;
        _items.RemoveAt(idx);
        OnRemoved(item, idx);
        return true;
    }

    /// <summary>Remove the item at <paramref name="index"/>.</summary>
    public void RemoveAt(int index)
    {
        T item = _items[index];
        _items.RemoveAt(index);
        OnRemoved(item, index);
    }

    /// <summary>Replace the item at <paramref name="index"/> with <paramref name="newItem"/>.</summary>
    public void Replace(int index, T newItem)
    {
        T oldItem = _items[index];
        _items[index] = newItem;
        OnReplaced(newItem, oldItem, index);
    }

    /// <summary>Remove all items and emit <see cref="Reset"/>.</summary>
    public void Clear()
    {
        _items.Clear();
        OnReset();
    }

    // ── Batch update ──────────────────────────────────────────────────────────

    /// <summary>
    /// Begin a batch-update scope. Granular events are suppressed until the
    /// returned token is disposed. On dispose of the outermost scope (ref-counted),
    /// a single <see cref="Reset"/> fires if any mutations occurred.
    /// </summary>
    public IDisposable BatchUpdate() => new BatchToken(this);

    private sealed class BatchToken : IDisposable
    {
        private readonly ObservableList<T> _owner;
        private bool _disposed;

        internal BatchToken(ObservableList<T> owner)
        {
            _owner = owner;
            if (_owner._batchDepth == 0)
                _owner._countAtBatchStart = _owner._items.Count;
            _owner._batchDepth++;
        }

        public void Dispose()
        {
            if (_disposed) return;
            _disposed = true;
            _owner._batchDepth--;
            if (_owner._batchDepth == 0 && _owner._mutatedInBatch)
            {
                int finalCount = _owner._items.Count;
                _owner._mutatedInBatch = false;
                _owner.Reset?.Invoke(_owner, EventArgs.Empty);
                _owner.CollectionChanged?.Invoke(
                    _owner,
                    new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Reset));
                // Emit Count notification only if count actually changed (spec §3.3).
                if (finalCount != _owner._countAtBatchStart)
                    _owner.PropertyChanged?.Invoke(_owner, new PropertyChangedEventArgs(nameof(Count)));
            }
        }
    }

    // ── Internal helpers ──────────────────────────────────────────────────────

    private void OnAdded(T item, int index)
    {
        if (_batchDepth > 0) { _mutatedInBatch = true; return; }

        ItemAdded?.Invoke(this, new ItemAddedEventArgs<T>(item, index));
        CollectionChanged?.Invoke(this, new NotifyCollectionChangedEventArgs(
            NotifyCollectionChangedAction.Add, item, index));
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(Count)));
    }

    private void OnRemoved(T item, int index)
    {
        if (_batchDepth > 0) { _mutatedInBatch = true; return; }

        ItemRemoved?.Invoke(this, new ItemRemovedEventArgs<T>(item, index));
        CollectionChanged?.Invoke(this, new NotifyCollectionChangedEventArgs(
            NotifyCollectionChangedAction.Remove, item, index));
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(Count)));
    }

    private void OnReplaced(T newItem, T oldItem, int index)
    {
        if (_batchDepth > 0) { _mutatedInBatch = true; return; }

        ItemReplaced?.Invoke(this, new ItemReplacedEventArgs<T>(newItem, oldItem, index));
        CollectionChanged?.Invoke(this, new NotifyCollectionChangedEventArgs(
            NotifyCollectionChangedAction.Replace, newItem, oldItem, index));
        // Count does not change on replace — no PropertyChanged("Count")
    }

    private void OnReset()
    {
        if (_batchDepth > 0) { _mutatedInBatch = true; return; }

        Reset?.Invoke(this, EventArgs.Empty);
        CollectionChanged?.Invoke(this, new NotifyCollectionChangedEventArgs(
            NotifyCollectionChangedAction.Reset));
    }
}
